### PR TITLE
[apex] Fix parsing of triggers with declarations

### LINKED ISSUE #5139: [apex] OperationWithHighCostInLoop not firing in triggers
**Affects PMD Version:** 7.3.0

**Rule:** [OperationWithHighCostInLoop](https://docs.pmd-code.org/latest/pmd_rules_apex_performance.html#operationwithhighcostinloop)

**Description:**
Violations for `OperationWithHighCostInLoop` no longer happen within triggers.

**Code Sample demonstrating the issue:**

```apex
trigger accountTrigger on Account (before insert, before update) {
		integer i = 0;
		for (i = 0; i <15; i++) {
			SObjectType token = Schema.getGlobalDescribe().get('Account');
		}
}
```

**Expected outcome:**
A violation for `OperationWithHighCostInLoop` should be thrown, the same way it was in 7.2.0.

**Running PMD through:** *[CLI]*

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

Only the grandchildren of a trigger block were ending up in the tree, but the direct children of triggerBlock were missing, e.g. ForLoopStatement. This caused OperationWithHighCostInLoop to not find the loop anymore in triggers.

This will probably fix other false negatives in triggers in other rules as well.

## Related issues

- Fixes #5139

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/design/xml/CyclomaticComplexity.xml [hunk 1]
```diff
@@ -263,7 +263,7 @@ trigger CaseAssignLevel on CaseAssignLevel__c (after delete, after insert, after
         <rule-property name="methodReportLevel">1</rule-property>
         <expected-problems>1</expected-problems>
         <expected-messages>
-            <message>The trigger 'CaseAssignLevel' has a cyclomatic complexity of 9.</message>
+            <message>The trigger 'CaseAssignLevel' has a cyclomatic complexity of 12.</message>
```
Comment (adangel) [inline pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/design/xml/CyclomaticComplexity.xml:266]: This actually undoes what I've done in https://github.com/pmd/pmd/commit/581b6db3d26627e2a7955352ef3cbfdf367d6e22 during the update of the apex-parser/summit-ast

**Comment by jsotuyod** [comment]
So, if I understand correctly, after the upgrade triggers were no longer an ASTUserClass, but a standalone and more specific ASTUserTrigger.

If this is the case, that should explain most if not all the differences in #5138, as it basically means triggers were not being analyzed in Java rules...

I wonder if any xpath rules were pointing directly to ASTUserClass and needs adjustments too...

**Comment by adangel** [comment]
It is still a ASTUserTrigger, but it didn't contain all statements...

e.g. compare the difference of the now existing test file for triggers between 7.2.0 and 7.3.0:

```
$ pmd-bin-7.2.0/bin/pmd ast-dump --file AccountTrigger.cls --language apex --format text
└─ ApexFile
   └─ UserTrigger
      ├─ ModifierNode
      └─ Method
         ├─ ModifierNode
         ├─ VariableDeclarationStatements
         │  ├─ ModifierNode
         │  └─ VariableDeclaration
         │     ├─ LiteralExpression
         │     └─ VariableExpression
         │        └─ EmptyReferenceExpression
         ├─ ForLoopStatement
         │  ├─ StandardCondition
         │  │  └─ BooleanExpression
         │  │     ├─ VariableExpression
         │  │     │  └─ EmptyReferenceExpression
         │  │     └─ LiteralExpression
         │  ├─ Expression
         │  │  └─ AssignmentExpression
         │  │     ├─ VariableExpression
         │  │     │  └─ EmptyReferenceExpression
         │  │     └─ LiteralExpression
         │  ├─ BlockStatement
         │  │  └─ VariableDeclarationStatements
         │  │     ├─ ModifierNode
         │  │     └─ VariableDeclaration
         │  │        ├─ MethodCallExpression
         │  │        │  ├─ ReferenceExpression
         │  │        │  │  └─ MethodCallExpression
         │  │        │  │     └─ ReferenceExpression
         │  │        │  └─ LiteralExpression
         │  │        └─ VariableExpression
         │  │           └─ EmptyReferenceExpression
         │  └─ PostfixExpression
         │     └─ VariableExpression
         │        └─ EmptyReferenceExpression
         ├─ VariableDeclarationStatements
         │  ├─ ModifierNode
         │  └─ VariableDeclaration
         │     ├─ LiteralExpression
         │     └─ VariableExpression
         │        └─ EmptyReferenceExpression
         └─ ExpressionStatement
            └─ MethodCallExpression
               ├─ ReferenceExpression
               └─ LiteralExpression
```

and

```
$ pmd-bin-7.3.0/bin/pmd ast-dump --file AccountTrigger.cls --language apex --format text
└─ ApexFile
   └─ UserTrigger
      ├─ ModifierNode
      └─ Method
         ├─ ModifierNode
         ├─ FieldDeclaration
         │  ├─ LiteralExpression
         │  └─ VariableExpression
         │     └─ EmptyReferenceExpression
         ├─ AssignmentExpression
         │  ├─ VariableExpression
         │  │  └─ EmptyReferenceExpression
         │  └─ LiteralExpression
         ├─ BooleanExpression
         │  ├─ VariableExpression
         │  │  └─ EmptyReferenceExpression
         │  └─ LiteralExpression
         ├─ BlockStatement
         │  └─ VariableDeclarationStatements
         │     ├─ ModifierNode
         │     └─ VariableDeclaration
         │        ├─ MethodCallExpression
         │        │  ├─ ReferenceExpression
         │        │  │  └─ MethodCallExpression
         │        │  │     └─ ReferenceExpression
         │        │  └─ LiteralExpression
         │        └─ VariableExpression
         │           └─ EmptyReferenceExpression
         ├─ PostfixExpression
         │  └─ VariableExpression
         │     └─ EmptyReferenceExpression
         ├─ FieldDeclaration
         │  ├─ LiteralExpression
         │  └─ VariableExpression
         │     └─ EmptyReferenceExpression
         └─ MethodCallExpression
            ├─ ReferenceExpression
            └─ LiteralExpression
```

e.g. the for-loop-statement is missing. only some children of the for-loop are there (some assignment, booleanexpression, blockstatement, ...)

The fix was basically using `buildAndSetParent` instead of `buildChildren` - the latter only as the children, but not the node itself (which was the among others the for-loop-statement).

I compared the tree also against 6.55.0 and figured, that the "Field" nodes should be added to UserTrigger and not the synthetic "invoke" method (which Jorje created and we do as well).

The change in the grammar is however: Previously (in 7.2.0/7.3.0) the first line in the trigger body (`integer i = 0;`) was parsed as VariableDeclarationStatements, but is now parsed as FieldDeclarationStatements. And when there are FieldDeclaration, Jorje adds these again as "Field" nodes.

**Comment by adangel** [comment]
And for reference, in 6.55.0 it looked like this:

```
$ pmd-bin-6.55.0/bin/run.sh ast-dump --file AccountTrigger.cls --language apex --format text
-------------------------------------------------------------------------------
This command line utility is experimental. It might change at any time without
prior notice.
-------------------------------------------------------------------------------
└─ UserTrigger
   ├─ ModifierNode
   ├─ Field
   │  └─ ModifierNode
   ├─ Field
   │  └─ ModifierNode
   ├─ Method
   │  ├─ ModifierNode
   │  ├─ FieldDeclarationStatements
   │  │  ├─ ModifierNode
   │  │  └─ FieldDeclaration
   │  │     ├─ LiteralExpression
   │  │     └─ VariableExpression
   │  │        └─ EmptyReferenceExpression
   │  ├─ ForLoopStatement
   │  │  ├─ ExpressionStatement
   │  │  │  └─ AssignmentExpression
   │  │  │     ├─ VariableExpression
   │  │  │     │  └─ EmptyReferenceExpression
   │  │  │     └─ LiteralExpression
   │  │  ├─ StandardCondition
   │  │  │  └─ BooleanExpression
   │  │  │     ├─ VariableExpression
   │  │  │     │  └─ EmptyReferenceExpression
   │  │  │     └─ LiteralExpression
   │  │  ├─ PostfixExpression
   │  │  │  └─ VariableExpression
   │  │  │     └─ EmptyReferenceExpression
   │  │  └─ BlockStatement
   │  │     └─ VariableDeclarationStatements
   │  │        ├─ ModifierNode
   │  │        └─ VariableDeclaration
   │  │           ├─ MethodCallExpression
   │  │           │  ├─ ReferenceExpression
   │  │           │  │  └─ MethodCallExpression
   │  │           │  │     └─ ReferenceExpression
   │  │           │  └─ LiteralExpression
   │  │           └─ VariableExpression
   │  │              └─ EmptyReferenceExpression
   │  ├─ FieldDeclarationStatements
   │  │  ├─ ModifierNode
   │  │  └─ FieldDeclaration
   │  │     ├─ LiteralExpression
   │  │     └─ VariableExpression
   │  │        └─ EmptyReferenceExpression
   │  └─ ExpressionStatement
   │     └─ MethodCallExpression
   │        ├─ ReferenceExpression
   │        └─ LiteralExpression
   └─ Method
      └─ ModifierNode
```

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/10125542991?pr=5146">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**